src/lsm/cache_map.zig

```diff
@@ -282,15 +280,12 @@ pub fn CacheMapType(
 
                     // A tombstone in the rollback log can only occur when the value doesn't exist
                     // in _both_ the cache and stash on insert.
-                    if (self.cache) |*cache| {
-                        // If we have cache enabled, it must be there.
-                        const cache_removed = cache.remove(key) != null;
-                        assert(cache_removed);
-                    }
```

Comment:
Ahhh, so I actually had a wrong mental model for the cache map! Rollbacks _dont_ guarantee that it ends up in _physically_ the same state, only that it would be logically the same. 

After the insert, the value here was in cache. But, before the rollback, it is in stash, not in cache. 

🤔 

I am now wondering if we compute the size of the stash correctly? That is, when we roll-forward, we spill some values into the stash. But when we roll back, because we are doing upserts, we might spill _more_ values into the stash?

That is, inserting `N` values into cache map does `N` inserts into the stash, but subsequent roll-back can do N more. I think this should be fine, as the keys are the same, but we'll need to double-check this!

Reply:
Stepping through the logic:

During scope rollback, each stash insert comes from (non-tombstone) values in the scope rollback log.

Each of those inserts in the rollback log corresponds to either:
- a stash remove during scope roll-forward (capacity for the insert in the stash is guaranteed by the remove), or
- a cache eviction during scope roll-forward (capacity for the insert in the stash is unused).

So I think the stash capacity is correct.

Reply:
>a stash remove during scope roll-forward

Does that leave a (hashmap-level) tombstone in place which still uses up capacity?

Reply:
Hashmap-level tombstones don't contribute to hashmap capacity, do they?

Reply:
Ah, right, tombstones are always re-used. I thought that might not be the case, but, checking the code and writing the test confirms that this is in fact fine!